### ISSUE TITLE
Add test coverage automation and reporting

### ISSUE THREAD

**Issue by rdoursenaud** [mido/mido#557]

**Comment by akx** [comment]
https://github.com/python-babel/babel/pull/833 might help here :)

**Comment by etienne-mad** [comment]
Hi @rdoursenaud,

I'd be happy to take this on if you're open to contributions. I have a bit of experience setting up tests and test coverage pipelines. Please let me know if you're interested

**Comment by rdoursenaud** [comment]
Yes, of course contributions of all kind are greatly appreciated. I haven't been able to do as much as I hoped due to personnal ongoing issues. Go ahead ! Thanks. Feel free to ping me if you have any inquiries.

**Comment by etienne-mad** [comment]
Thanks for the go-ahead on this.

Before writing any code, I want to make sure the scope matches what you had in mind and get your input on a few decisions.

---

I ran coverage locally after setting up the dev environment. Two numbers worth knowing:

- `coverage run -m pytest`: **76%** — naive run, no source filter. Biased upward because `tests/*` is included in the measurement.
- `coverage run --source=mido -m pytest`: **47%** — honest baseline. Also reveals 4 CLI scripts at 0% that are invisible without an explicit source filter.

---

**Proposed scope**

In this PR:
- `pyproject.toml`: new `test-coverage` dependency group, coverage configuration (`[tool.coverage.run]` and `[tool.coverage.report]`)
- `.github/workflows/test.yml`: new `coverage` job, HTML report uploaded as workflow artifact

Deferred to follow-up PRs:
- Multi-OS / multi-Python coverage with `coverage combine` across the matrix
- `fail_under` threshold enforcement (needs a baseline observation period first)
- Integration with external services (Codecov, Coveralls, etc.) — account, token, and governance decisions that should be yours to make separately

Separate issues:
- New tests to improve coverage of specific modules
- One flaky test I noticed during setup (time-based assertion) — I'll file it separately

---

**Questions**

**1 — Scope alignment**
Does the scope above match what you had in mind, or would you like me to narrow or widen it?

**2 — `dev` group composition**
The new `test-coverage` group will include `test-code` through group composition (`mido[test-code]`). For the existing `dev` group, should I (a) replace `mido[test-code]` with `mido[test-coverage]`, or (b) list both side by side? I recommend **(b)** for explicitness.

**3 — Coverage job in the matrix**
The current `test` job runs 7 Python × 3 OS = 21 combinations. I'd like to add a `coverage` job on a single combination (ubuntu-latest + Python 3.12). Rationale: the low-coverage modules are hardware/network constrained and don't vary meaningfully by OS. Does this work?

---

**Decisions already made (no action needed)**

- Native GitHub reporting only: terminal summary + XML + HTML artifact via `actions/upload-artifact@v4`. No external service — account and token decisions should be yours to make separately.
- Line coverage only (`branch = false`) to start. Keeps the baseline comparable. One-line change to enable later.
- No `omit` of low-coverage modules. Hiding them would inflate the number and obscure a reality the project already accepts.

---

Ready to code once I have your feedback.

**Comment by rdoursenaud** [comment]
> Thanks for the go-ahead on this.

You're welcome, thanks for chiming in!

> Before writing any code, I want to make sure the scope matches what you had in mind and get your input on a few decisions.

Thank you, very much appreciated.

> I ran coverage locally after setting up the dev environment. Two numbers worth knowing:
> 
> * `coverage run -m pytest`: **76%** — naive run, no source filter. Biased upward because `tests/*` is included in the measurement.
> * `coverage run --source=mido -m pytest`: **47%** — honest baseline. Also reveals 4 CLI scripts at 0% that are invisible without an explicit source filter.

Not surprising to me. The scripts are meant to be mere examples of how to (naively) use the library. Covering them is not high priority but could reveal the need to update them when library modifications break them.

> **Proposed scope**
> 
> In this PR:
> 
> * `pyproject.toml`: new `test-coverage` dependency group, coverage configuration (`[tool.coverage.run]` and `[tool.coverage.report]`)
> * `.github/workflows/test.yml`: new `coverage` job, HTML report uploaded as workflow artifact

Excellent!

> Deferred to follow-up PRs:
> 
> * Multi-OS / multi-Python coverage with `coverage combine` across the matrix
> * `fail_under` threshold enforcement (needs a baseline observation period first)
> * Integration with external services (Codecov, Coveralls, etc.) — account, token, and governance decisions that should be yours to make separately

Personally, I'd rather avoid external dependencies as much as possible. However it's up for debate with the community and if these provide significant upsides, I'd be willing to provide the appropriate accounts and maintain proper GitHub Actions secrets.

> Separate issues:
> 
> * New tests to improve coverage of specific modules
> * One flaky test I noticed during setup (time-based assertion) — I'll file it separately

I've already identified that one test as very problematic. It fails on low-end machines or if the test infrastructure is highly loaded. It was important to validate the benefits of an optimization but could probably be rewritten to be more predictable. I can dig the related issue and PR history if need be (A git blame on the test should already give some insights).

> **Questions**
> 
> **1 — Scope alignment** Does the scope above match what you had in mind, or would you like me to narrow or widen it?

Looks good to me. You may want to have a look at #540 for what I have in mind in the long run and some interesting exchanges with @akx. 

> **2 — `dev` group composition** The new `test-coverage` group will include `test-code` through group composition (`mido[test-code]`). For the existing `dev` group, should I (a) replace `mido[test-code]` with `mido[test-coverage]`, or (b) list both side by side? I recommend **(b)** for explicitness.

I also prefer it to be explicit. Lines of code are relatively cheap, human understanding is not ;)

> **3 — Coverage job in the matrix** The current `test` job runs 7 Python × 3 OS = 21 combinations. I'd like to add a `coverage` job on a single combination (ubuntu-latest + Python 3.12). Rationale: the low-coverage modules are hardware/network constrained and don't vary meaningfully by OS. Does this work?

Yeah, I've gone a bit overkill on the matrix to make sure we don't break any officially supported versions. Coverage however is a different beast and IMHO only needs to signal to maintainers/developers the current state of affairs. No need to go for all Python versions, only the latest supported. OS support however can vary and we plan to someday add native backends. Your proposed solution is good for now but I'd like to at least have every single OS covered sometimes in the future. Could we maybe restrict specific OS coverage to a single module? I.e. no need to test ALSA backend on Microsoft Windows and MacOS and so on.

> **Decisions already made (no action needed)**
> 
> * Native GitHub reporting only: terminal summary + XML + HTML artifact via `actions/upload-artifact@v4`. No external service — account and token decisions should be yours to make separately.
> * Line coverage only (`branch = false`) to start. Keeps the baseline comparable. One-line change to enable later.
> * No `omit` of low-coverage modules. Hiding them would inflate the number and obscure a reality the project already accepts.

Perfect, no notes!

> Ready to code once I have your feedback.

Again, thank you for tackling this.
